ui/lib/utils/tableColumnFactory.tsx

```diff
@@ -229,26 +225,64 @@ export class BarColumn {
     return this.factory.singleBar(fieldName, unit, rows)
   }
 
-  multiple<T>(bars: BarsConfig<T>, unit: string, rows?: T[]) {
+  multiple<T>(bars: DerivedBar, unit: string, rows?: T[]) {
     return this.factory.multipleBar(bars, unit, rows)
   }
 }
 
 ////////////////////////////////////////////
 
+export type DerivedFields = Record<
+  string,
+  DerivedBar['sources'] | DerivedCol['sources']
+>
+
+export function genDerivedBarSources(
+  avg: string,
+  max: string,
+  min?: string
+): DerivedBar['sources'] {
+  const res = [
+    {
+      tooltipPrefix: 'mean',
+      fieldName: avg,
+    },
+    {
+      tooltipPrefix: 'max',
+      fieldName: max,
+    },
+  ]
+  if (min) {
+    res.push({
+      tooltipPrefix: 'min',
+      fieldName: min,
+    })
+  }
+  return res
+}
+
+function isDerivedBarSources(v: any): v is DerivedBar['sources'] {
+  return !!v[0].fieldName
+}
+
 export function getSelectedFields(
   visibleColumnKeys: IColumnKeys,
-  columns: IColumnWithSourceFields[]
+  derivedFields: DerivedFields
 ) {
   let fields: string[] = []
-  columns.forEach((c) => {
-    if (visibleColumnKeys[c.key] === true) {
-      if (c.sourceFields !== undefined) {
-        fields = fields.concat(c.sourceFields)
+  let sources: DerivedFields[keyof DerivedFields]
+  for (const columnKey in visibleColumnKeys) {
+    if (visibleColumnKeys[columnKey]) {
+      if ((sources = derivedFields[columnKey])) {
```

Comment:
Curious what's this kind of usage? `((...))`

Reply:
just to avoid lint reporting warnings